### ISSUE TITLE
torch.rand() skewed for large tensors

### ISSUE THREAD

**Issue by meganset** [pytorch/pytorch#27946]
## 🐛 Bug

```
>>> torch.__version__
'1.3.0'
>>> torch.rand(6000000000).mean()
tensor(0.5000)
>>> torch.rand(6500000000).mean()
tensor(0.4956)
>>> torch.rand(8000000000).mean()
tensor(0.4027)
```

**Comment by albanD** [comment]
Hi,

I cannot reproduce that locally:
```
In [1]: import torch

In [2]: torch.rand(6000000000).mean()
Out[2]: tensor(0.5000)

In [3]: torch.rand(8000000000).mean()
Out[3]: tensor(0.5000)

In [4]: torch.__version__
Out[4]: '1.3.0a0+477f474'
```

Also this could be accumulation errors, what does the following return on your system: `torch.rand(8000000000).double().mean()` ?

**Comment by meganset** [comment]
sorry, I don't have enough memory to try the mean in double precision on the 8 billion-element tensor.
64g total memory on my linux box.

I see the skew on somewhat smaller tensors over on an imac with 32g memory, 
but the mean looks good in double precision.
e.g.

```
>>> torch.rand(3000000000).mean()
tensor(0.1790)
>>> torch.rand(3000000000).double().mean()
tensor(0.5000, dtype=torch.float64)

```

**Comment by albanD** [comment]
I would guess that the mean does `t.sum().div(t.nelements())` and so if you have too many elements, you can loose precision when doing this.

**Comment by vishwakftw** [comment]
@meganset could you try `torch.randn(3000000000).sum()` and check if it is close to `1500000000`?

**Comment by albanD** [comment]
I would expect `torch.randn()` not to suffer from the same issue, even for `torch.randn(8000000000).mean()` as the sum will remain close to 0 and so there won't be precision loss.

**Comment by vishwakftw** [comment]
Oh sorry, I meant `torch.rand`.

**Comment by meganset** [comment]
Thanks for looking at this.
I retreated to c++ to verify the loss of precision in any long sum of 4-byte floats:

```
void f(int64_t n) {  // length, in millions
 n*=1000000;
 auto t=torch::rand(n);
 double d=0; float f=0,*p=t.data_ptr<float>();
 for(size_t i=0; i<n; ++i) d+=p[i], f+=p[i];
 std::cerr << "double  sum: " <<   d << "\n";
 std::cerr << "float   sum: " <<   f << "\n\n";
 std::cerr << "double mean: " << d/n << "\n";
 std::cerr << "float  mean: " << f/n << "\n";
 std::cerr << "torch  mean: " << t.mean().item().toFloat() << "\n";
}
```
the loss of precision in the sum begins to occur around 33 million elements,
noticeable by around 40 million, e.g.

```
f(33);
double  sum: 1.64992e+07
float   sum: 1.65002e+07

double mean: 0.499975
float  mean: 0.500007
torch  mean: 0.499975

f(40);
double  sum: 2.00009e+07
float   sum: 1.67772e+07

double mean: 0.500021
float  mean: 0.41943   <<<
torch  mean: 0.500022
```
But the torch _t.mean()_ is still correct, I assume it is using double precision in the sum.
The torch calculated mean will not go astray on my linux box until I try a tensor of around 14 billion elements:

```
f(14000);
double  sum: 7.00001e+09
float   sum: 1.67772e+07

double mean: 0.500001
float  mean: 0.00119837
torch  mean: 0.460175       <<<
```
This boundary is a little mysterious, as it would seem that the double precision accumulation is still arriving at the correct sum & mean.
Also mysterious is why the python version starts to go wrong earlier:
on my linux box, the mean is off when I use _torch.rand(8 billion)_
But not on albanD's box..?

**Comment by albanD** [comment]
Interestingly, to speed up testing, I did the following and I see the same results as you in this case:

```
In [1]: import torch

In [2]: a = torch.zeros(8000000000).add_(0.5)

In [3]: a.mean()
Out[3]: tensor(0.4027)

In [4]: a.narrow(0, 0, 6500000000).mean()
Out[4]: tensor(0.4956)

In [5]: a.narrow(0, 0, 6000000000).mean()
Out[5]: tensor(0.5000)
```

> This boundary is a little mysterious, as it would seem that the double precision accumulation is still arriving at the correct sum & mean.

The answer here is that the mean is actually super fast (<1s) and so you probably don't see it but it uses multithreading ;)

```
In [1]: import torch

In [2]: a = torch.zeros(int(1e9)).add_(0.5)

In [3]: a.mean()
Out[3]: tensor(0.5000)

In [4]: torch.get_num_threads()
Out[4]: 12

In [5]: torch.set_num_threads(1)

In [6]: a.mean()
Out[6]: tensor(0.2684)
```

Basically, when the workload is big enough, we use multithreading. This has the side effect to perform several sums and then accumulate them. Leading to extra precision. This is why the precision will be different depending on the number of processors you have.

**Comment by meganset** [comment]
Ah.
I see now.
Thanks.

My issue title is misleading, should be something like
"Summary statistics on large float tensors lose precision depending on number of threads used"

This also explains how c++ might differ from python if they use a different number of threads.
I guess it's still a bug though, to get a different mean depending on how many threads are available.
And _t.mean()_ should still be able to sum & divide in double precision.

The _torch.mean()_ function anticpates this by allowing a dtype=torch.double option,
but this seems to convert the whole tensor to double, rather than just the sum

**Comment by soumith** [comment]
I'm going to argue that this is expected behavior as well. We've always given guidance that float32 is imprecise and is expected to be so.

**Comment by albanD** [comment]
I'm going to close base on the comment.
Feel free to re-open if you think there is something more we could do here.
